### PR TITLE
feat: Update state congruence [PAGOPA-1718]

### PR DESCRIPTION
**Pull request by AngeloCaporaso**
#### List of Changes

- Update status sync with API calls: 
   - when `currentDate` > `expirationDate` && otherConditionsToExpire -> set status to `EXPIRED`
   - when `currentDate` > `validityDate` && otherConditionsToExpire -> set status to `VALID`

This update is made when an `PaymentPosition` receives requests for operations including:
- `pay`
- `report`
- `invalidate`

`Create` and `Update` request are excluded because they already include the `PaymentPosition` logic for validity and expiration.

#### Motivation and Context

To date, state changes are based on batches. If the batch frequency is N, between times t-N and t the required state changes based on validityDate and expirationDate are lost, i.e., postponed. One solution would be to decrease the batch frequency to a relatively small time, e.g., 1s, but this could pose a performance risk, so we choose to update state on the fly in memory only when necessary.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] My change doesn't require a change in documentation.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/debtposition/model/enumeration/DebtPositionStatus.java [hunk 1]
```diff
@@ -37,4 +41,18 @@ public static Set<DebtPositionStatus> getPaymentPosFullyPaidStatus() {
     public static Set<DebtPositionStatus> getPaymentPosNotAccountableStatus() {
         return EnumSet.of(DRAFT, PUBLISHED, VALID, INVALID, EXPIRED, REPORTED);
     }
+
+    public static PaymentPosition updatePaymentPositionStatus(PaymentPosition pp) {
+        LocalDateTime currentDate = LocalDateTime.now(ZoneOffset.UTC);
+        // Validity check on the fly
+        if(pp.getStatus().equals(DebtPositionStatus.PUBLISHED) && currentDate.isAfter(pp.getValidityDate())) {
```
Comment (FedericoRuzzier) [inline src/main/java/it/gov/pagopa/debtposition/model/enumeration/DebtPositionStatus.java:48]: if validityDate is null there can be a nullPointerException

File: src/main/java/it/gov/pagopa/debtposition/model/enumeration/DebtPositionStatus.java [hunk 2]
```diff
@@ -37,4 +41,18 @@ public static Set<DebtPositionStatus> getPaymentPosFullyPaidStatus() {
     public static Set<DebtPositionStatus> getPaymentPosNotAccountableStatus() {
         return EnumSet.of(DRAFT, PUBLISHED, VALID, INVALID, EXPIRED, REPORTED);
     }
+
+    public static PaymentPosition updatePaymentPositionStatus(PaymentPosition pp) {
```
Comment (FedericoRuzzier) [inline src/main/java/it/gov/pagopa/debtposition/model/enumeration/DebtPositionStatus.java:45]: Can be void?

Comment (AngeloCaporaso) [reply src/main/java/it/gov/pagopa/debtposition/model/enumeration/DebtPositionStatus.java:45]: I prefer that the modifier method returns the modified object, to make the code more readable and fluent ([ref](https://www.baeldung.com/java-fluent-interface-vs-builder-pattern#fluent-interface)). Also, I think that in theory, by keeping this interface, you could also return a copy of the object if necessary.

**Review by FedericoRuzzier — APPROVED** [review]
LGTM 💯

**Pull request merged**